server/src/main/java/org/elasticsearch/common/file/AbstractFileWatchingService.java

```diff
@@ -116,15 +116,13 @@ public final boolean watching() {
     }
 
     private FileUpdateState readFileUpdateState(Path path) throws IOException {
-        BasicFileAttributes attr;
         try {
-            attr = filesReadAttributes(path, BasicFileAttributes.class);
+            BasicFileAttributes attr = filesReadAttributes(path, BasicFileAttributes.class);
+            return new FileUpdateState(attr.lastModifiedTime().toMillis(), path.toRealPath().toString(), attr.fileKey());
```

Comment:
`path.toRealPath()` can also throw `NoSuchFileException` and lead to failure like [this one](https://gradle-enterprise.elastic.co/s/oiobnexiergym/tests/task/:server:internalClusterTest/details/org.elasticsearch.snapshots.SharedClusterSnapshotRestoreIT/testDeletionOfFailingToRecoverIndexShouldStopRestore?top-execution=1). So moving it inside the try-catch.